influxdb3_server/src/http.rs

```diff
@@ -203,119 +221,32 @@ where
+    }
 
-        fn to_pretty(batches: Vec<RecordBatch>) -> Result<Bytes> {
-            Ok(Bytes::from(format!(
-                "{}",
-                pretty::pretty_format_batches(&batches)?
-            )))
-        }
+    async fn query_influxql(&self, req: Request<Body>) -> Result<Response<Body>> {
+        self.query_inner(req, QueryKind::InfluxQl).await
+    }
 
-        fn to_parquet(batches: Vec<RecordBatch>) -> Result<Bytes> {
-            let mut bytes = Vec::new();
-            let mem_pool = Arc::new(UnboundedMemoryPool::default());
-            let mut writer =
-                TrackedMemoryArrowWriter::try_new(&mut bytes, batches[0].schema(), mem_pool)?;
-            for batch in batches {
-                writer.write(batch)?;
-            }
-            writer.close()?;
-            Ok(Bytes::from(bytes))
-        }
+    async fn query_inner(&self, req: Request<Body>, kind: QueryKind) -> Result<Response<Body>> {
+        let QueryParams {
+            database,
+            query_str,
+            format,
+        } = QueryParams::from_request(&req)?;
 
-        enum Format {
-            Parquet,
-            Csv,
-            Pretty,
-            Json,
-            Error,
-        }
+        info!(%database, %query_str, ?format, ?kind, "handling query");
 
-        let (body, format) = match params.format {
-            None => match req
-                .headers()
-                .get(ACCEPT)
-                .map(HeaderValue::to_str)
-                .transpose()?
-            {
-                // Accept Headers use the MIME types maintained by IANA here:
-                // https://www.iana.org/assignments/media-types/media-types.xhtml
-                // Note parquet hasn't been accepted yet just Arrow, but there
-                // is the possibility it will be:
-                // https://issues.apache.org/jira/browse/PARQUET-1889
-                Some("application/vnd.apache.parquet") => {
-                    (to_parquet(batches)?, Format::Parquet)
-                }
-                Some("text/csv") => (to_csv(batches)?, Format::Csv),
-                Some("text/plain") => (to_pretty(batches)?, Format::Pretty),
-                Some("application/json") => (to_json(batches)?, Format::Json),
-                Some("*/*") | None => (to_json(batches)?, Format::Json),
-                Some(_) => (Bytes::from("{ \"error\": \"Available mime types are: application/vnd.apache.parquet, text/csv, text/plain, and application/json\" }"), Format::Error),
-            },
-            Some(format) => match format.as_str() {
-                "parquet" => (to_parquet(batches)?, Format::Parquet),
-                "csv" => (to_csv(batches)?, Format::Csv),
-                "pretty" => (to_pretty(batches)?, Format::Pretty),
-                "json" => (to_json(batches)?, Format::Json),
-                _ => (Bytes::from("{ \"error\": \"Available formats are: parquet, csv, pretty, and json\" }"), Format::Error),
```

Comment:
I made it so this error is handled when the request is first decomposed, so that a 400 is returned before doing any query processing.

So, the `QueryParams::<QueryFormat>::from_request` will emit this error now.

Reply:
Thanks for doing that. This was pre sending back errors via the error type for the server so glad you moved it out.